inject/hk2/src/main/java/org/glassfish/jersey/inject/hk2/SupplierFactoryBridge.java

```diff
@@ -74,7 +74,9 @@ public T provide() {
             Supplier<T> supplier = locator.getService(beanType, beanName);
             T instance = supplier.get();
             if (disposable) {
-                disposableSuppliers.put(instance, (DisposableSupplier<T>) supplier);
+                synchronized (disposableSuppliers) {
```

Comment:
Do not use synchronized block, rather create a synchronized map in field declaration:
`private Map<Object, DisposableSupplier<T>> disposableSuppliers = Collections.synchronizedMap(new IdentityHashMap<>());`